cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/utilities/assert.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/equal.h>
+#include <thrust/swap.h>
+#include <thrust/transform_reduce.h>
+
+#include <cuda/std/tuple>
+#include <cuda/std/utility>
+
+#include <limits>
+#include <memory>
+#include <utility>
+
+namespace cudf {
+namespace experimental {
+
+/**
+ * @brief A map from cudf::type_id to cudf type that excludes LIST and STRUCT types.
+ *
+ * To be used with type_dispatcher in place of the default map, when it is required that STRUCT and
+ * LIST map to void. This is useful when we want to avoid recursion in a functor. For example, in
+ * element_comparator, we have a specialization for STRUCT but the type_dispatcher in it is only
+ * used to dispatch to the same functor for non-nested types. Even when we're guaranteed to not have
+ * non-nested types at that point, the compiler doesn't know this and would try to create recursive
+ * code which is very slow.
+ *
+ * Usage:
+ * @code
+ * type_dispatcher<dispatch_nested_to_void>(data_type(), functor{});
+ * @endcode
+ */
+template <cudf::type_id t>
+struct dispatch_void_if_nested {
+  using type = std::conditional_t<cudf::is_nested(data_type(t)), void, id_to_type<t>>;
+};
+
+namespace row {
+
+namespace lexicographic {
+
+/**
+ * @brief Computes whether one row is lexicographically *less* than another row.
+ *
+ * Lexicographic ordering is determined by:
+ * - Two rows are compared element by element.
+ * - The first mismatching element defines which row is lexicographically less
+ * or greater than the other.
+ *
+ * Lexicographic ordering is exactly equivalent to doing an alphabetical sort of
+ * two words, for example, `aac` would be *less* than (or precede) `abb`. The
+ * second letter in both words is the first non-equal letter, and `a < b`, thus
+ * `aac < abb`.
+ *
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <typename Nullate>
+class device_row_comparator {
```

Comment:
Adding `element_comparator` as a nested type made me realize that `device_row_comparator` isn't publicly constructible either. That makes me think we should nest it inside of `self_comparator`. Especially since we'll need to add a different version for the non-self comparator.

Reply:
I purposefully left it out because then it can be used by both the one table comparator and the two table comparator. The one table comparator will use it directly while the two table comparator can wrap it in another index side aware class. That index side aware wrapper class can live inside two table comparator.